src/lando/pushlog/models/commit.py

```diff
@@ -239,13 +239,52 @@ class Tag(models.Model):
         on_delete=models.SET_NULL,
     )
 
+    _scm_commit: CommitData | None = None
+
+    @staticmethod
+    def for_scm_commit(repo: Repo, name: str, scm_commit: CommitData):  # noqa: ANN205
+        """Create a partial Tag object, based on CommitData not present in the DB
+        yet."""
+        try:
+            # If a commit already exists in the DB, use that
+            commit = Commit.objects.get(repo=repo, hash=scm_commit.hash)
+            return Tag(repo=repo, name=name, commit=commit)
+        except Commit.DoesNotExist:
+            pass
+
+        tag = Tag(repo=repo, name=name)
+        tag._scm_commit = scm_commit
+
+        return tag
+
+    def save(self, *args, **kwargs):
+        """Save the Tag data to the DB.
+
+        If the target commit is not an ORM model yet, this method will find or create
+        it as needed, and maintain the DB relations.
+        """
+        # First make sure we don't deal with stale data.
+        try:
+            self.refresh_from_db()
+        except Tag.DoesNotExist:
+            # We're OK if this commit doesn't exist in the DB yet; we're just about to
+            # write it.
+            pass
```

Comment:
This is a little strange here. What's the purpose of this? I.e., what data could be stale?

I think there might be some unexpected behaviour in the future, since it will revert any edits to the instance before saving it. Regardless, better to check for `self.pk` than to catch an exception here.

Reply:
As we create the object in memory before committing it to the DB (so we have something we can create relations to [which won't work until we commit it to the DB anyway...]), the refresh serves to fetch the latest version (which there shouldn't be because we only have a single worker processing a single job at any time), _before_ we make any local modification further down the line.

So... Yeah, it does look like overkill, as we're enabling something that we don't do, and protecting against something that's structurally no gonna happen.